File: house-robber/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> v(n+1, 0);
        v[0] = 0;
        v[1] = nums[0];
        for (int i = 2; i < n; i++) {
            v[i] = max(nums[i-1] + v[i-2], v[i-1]);
        }
        if (n == 1) {
            return nums[0];
        }
        v[n] = max(nums[n-1] + v[n-2], v[n-1]);
        return v[n];
    }
};

File: house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        dp = [0 for i in range(len(nums) + 1)]
        N = len(nums)
        dp[N], dp[N - 1] = 0, nums[N - 1]

        for i in range(N-2, -1, -1):
            dp[i] = max(dp[i+1], dp[i+2] + nums[i])
        return max(dp)
